# ECE1513 - Winter 2020
## Assignment 2
### Yixiao Hong 1001311145

### Dataset Preparation

In [1]:
#import libs
import numpy as np
import pandas as pd

#import dataset
from sklearn.datasets import load_breast_cancer
breast_cancer = load_breast_cancer()

#Extract data
X = breast_cancer.data
#Extract targets
Y = breast_cancer.target

#Extract training data
train_X = X[:450]
#Extract training data
test_X = X[450:]
#Extract testing targets
train_Y = Y[:450]
#Extract testing targets
test_Y = Y[450:]

#display first 5 data features
print("First 5 data points (features)")
display(pd.DataFrame(X).head())
#display first 5 data targets
print("First 5 data points (targets)")
display(pd.DataFrame(Y).head())

First 5 data points (features)


0      1       2       3   ...      26      27      28       29
0  17.99  10.38  122.80  1001.0  ...  0.7119  0.2654  0.4601  0.11890
1  20.57  17.77  132.90  1326.0  ...  0.2416  0.1860  0.2750  0.08902
2  19.69  21.25  130.00  1203.0  ...  0.4504  0.2430  0.3613  0.08758
3  11.42  20.38   77.58   386.1  ...  0.6869  0.2575  0.6638  0.17300
4  20.29  14.34  135.10  1297.0  ...  0.4000  0.1625  0.2364  0.07678

[5 rows x 30 columns]

First 5 data points (targets)


0
0  0
1  0
2  0
3  0
4  0

### Problem2.2
###  (8 points) Implement this modified perceptron algorithm in Python and NumPy. Include a copy of your code in your submission for this assignment.

#### Implement the perceptron algorithm

In [0]:
#Define the perceptiron algorithm function
def perceptron(train_X,train_Y,Lr,max_iter):
  """
  Inputs:
  tranin_X(npArray): the train data features
  train_Y(npArray): the train data targets
  max_iter(int): the maximun iterations
  
  Outputs:
  w(npArray):the trained weight matrix
  b(float):the trained bias term
  """
  #correst the tratgets from data set, make all 0 to be -1
  train_Y = np.where(train_Y==0, -1, train_Y)
  #Set up initial value of the weight
  w = np.zeros(len(train_X[0]))
  b = 0
  #Start the training
  for _ in range(max_iter):
    #set up the update flag
    update_flag = False
    for Xi, ti in zip(train_X,train_Y):
      zi = np.dot(Xi.T, w) + b
      if float(zi*ti) <= 0:
        #needs to update w and b
        w = w + Lr*np.dot(Xi.T, ti)
        b = b + Lr*ti
        update_flag = True
    if not update_flag:
      print("The weights and bias term converged")
      break
  if update_flag:
    print("Exceeds the max_iter, training end")
  return(w,b)

#### Perform the training using the function developed above 

In [3]:
#train the classifier with train data
#here we set the max iteration be 10000
print("===>Start the training\n")
w,b = perceptron(train_X,train_Y,1,10000)
print("\n===>Train finished:")
print("\n==>Trained w=:\n",w,"\n\n==>Trained b=\n",b)

===>Start the training

Exceeds the max_iter, training end

===>Train finished:

==>Trained w=:
 [103003.70599992 -15846.22000007 185366.87999993  -6956.6
  -2724.49909    -13820.22614001 -19447.8850209   -7691.45253399
  -3562.4856      -1065.76898      2310.3296      10259.0284
 -19625.40059998 -17486.44999997   -316.339701    -3142.182471
  -4254.3529819    -899.798401    -1185.348957     -238.7286058
 104882.68199998 -53353.99999992 -96301.67999996  -3329.7
  -4763.45449    -44815.94142004 -54303.73955601 -14413.09809597
 -11790.29380001  -4332.38642999] 

==>Trained b=
 12938


#### Report the test error of your perceptron on the breast cancer dataset included with sklearn. Split the dataset (which includes 569 points) in 450 training examples and 119 test examples.

In [4]:
#use the trained classifier to predict the test examples
predict = np.dot(test_X,w.T) + b

print("==> predicted targets for the test data")
#correct the prediction result to binary term 1/0
predict = np.where(predict>=0, 1, predict)
predict = np.where(predict<0, 0, predict)
predict = predict.astype(int)
print(predict)

print("==> original true targets for the test data:")
print(test_Y)

==> predicted targets for the test data
[1 0 1 1 1 1 1 1 1 1 0 0 1 1 1 0 1 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 0 1 0 1 0 0 1 1 1 1 1 0 0 1 1 1 0 1 1 1 1 1 1 1 1 0 1 0 1 0 0 1 1 1 0 1 1
 1 1 1 1 1 1 1 1 1 0 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 0 0 0 0 0 0 1]
==> original true targets for the test data:
[1 0 1 1 1 1 1 1 1 1 0 0 1 1 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1 0 1 1 1 1 1 1 1
 0 1 0 1 1 0 1 1 1 1 1 0 0 1 0 1 0 1 1 1 1 1 0 1 1 0 1 0 1 0 0 1 1 1 0 1 1
 1 1 1 1 1 1 1 1 1 0 1 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 0 0 0 0 0 0 1]


#### Report the test error

In [5]:
#calculate the test error
from sklearn.metrics import confusion_matrix
tn, fp, fn, tp = confusion_matrix(test_Y, predict).ravel()
print("==>True Positive Rate (Recall)\n  = tp/(tp+fn):\n  =",round((tp/(tp+fn))*100,2),"%")
print("==>True Negative Rate (Selectivity)\n  = tn/(tn+fp):\n  =",round((tn/(tn+fp))*100,2),"%")

from sklearn.metrics import accuracy_score
print("==>Accuracy\n  = (tp+tn)/(tp+tn+fp+fn)\n  =",round((accuracy_score(test_Y, predict))*100,2),"%")


==>True Positive Rate (Recall)
  = tp/(tp+fn):
  = 97.83 %
==>True Negative Rate (Selectivity)
  = tn/(tn+fp):
  = 85.19 %
==>Accuracy
  = (tp+tn)/(tp+tn+fp+fn)
  = 94.96 %


#### The results appears to be good with True Positive Rate = 97.83% , True Negative Rate = 85.19% and Accuracy = 94.96%